proposals/awslogs.md

```diff
@@ -0,0 +1,195 @@
+# ECS CLI Logging
+
+## Overview
+
+The following Proposal lays out a design and implementation plan for creating a user experience in the ECS CLI for getting container logs from CloudWatch.
+
+### Use Cases
+
+1. User has a known error and wants to find more info on it.
+2. User is doing a deployment and wants to tail the logs and grep for errors.
+3. User wants to quickly set up their CloudWatch Logs based upon the configuration specified in their docker compose file, and have the CLI creates any necessary log groups for them.
+4. User wants to monitor their task/service, so they continually stream the logs.
+
+## Phase 1 Solution
+Top level `ecs-cli logs` command that will not use the docker compose file. This allows it to be used by a wide array of ECS customers, not just compose users. The command will allow customers to find logs for a given task.
+
+```
+ecs-cli logs --help
+--follow                         Stream logs (continuously poll for updates)
+--task-id                        [Required] View logs for a given Task ID
+--task-def                      Required with Task ID if the task has been stopped already. Format: family:revision
+--pattern                        Substring to search for within the logs.
+--container-name, -c     Filter logs for a given container definition
+--since                           Filter logs in the last X minutes (can not be used with start time and end time)
+--start-time                    Filter logs within a time frame, use with --end-time
+--end-time                     Filter logs within a time frame, use with --end-time
+--time-stamps, -t           View time-stamps with the logs
+```
+
+```
+ecs-cli logs --task-id d86079d1-6858-45e9-8ce2-1ba881c55c12 --time-stamps
+Time-stamp            Message
+2017-09-28 22:32:11 WordPress not found in /var/www/html - copying now...
+2017-09-28 22:32:11 Complete! WordPress has been successfully copied to /var/www/html
+2017-09-28 22:32:12 AH00558: apache2: Could not reliably determine the server's fully qualified domain name, using 172.17.0.3. Set the 'ServerName' directive globally to suppress this message
+2017-09-28 22:32:12 AH00558: apache2: Could not reliably determine the server's fully qualified domain name, using 172.17.0.3. Set the 'ServerName' directive globally to suppress this message
+2017-09-28 22:32:12 [Wed Sep 27 22:32:12.300422 2017] [mpm_prefork:notice] [pid 1] AH00163: Apache/2.4.10 (Debian) PHP/5.6.31 configured -- resuming normal operations
+2017-09-28 22:32:12 [Wed Sep 27 22:32:12.300456 2017] [core:notice] [pid 1] AH00094: Command line: 'apache2 -D FOREGROUND'
+```
+
+### Implementation
+
+- The logs implementation will not include any pagination- the command will return all logs corresponding to the specified search. We expect most users will be piping the output of the command to save it to a file, so this should not be a problem.
```

Comment:
> Why do we expect that customers will be piping to save it to a file rather than just viewing it in their shell? We can detect whether output is going to a pty/tty or a file.

Are you suggesting that we paginate the output if its sent to the command line?

> It might also be interesting to have different defaults for running and stopped tasks; we might want to follow running logs but not follow logs for stopped tasks.

I'd rather default to not following the logs in both cases; however, I'm now thinking that may using the `--follow` option on a stopped task should either result in an error- or it should print all the logs and then quit (so ignore the follow).

Reply:
I think if running `--follow` on a running task, and it is marked `essential` and becomes `STOPPED`, it should stop following.